js-plot/README.md

```diff
@@ -2,22 +2,214 @@
+  double x = initial_guess;
+  double delta_x = 0.0;
+  int i = 0;
+  do {
+    delta_x = equation(x) / derivative(x);
+    iterations.push_back({i++, x, equation(x), derivative(x), delta_x});
+    x = x - delta_x;
+  } while (fabs(delta_x) >= tolerance);
+  return x;
+};
+```
+File: _newtonraphson.cpp_.
+
+Before we go into Emscripten world, lets test if the iteration property is actually populated correctly by wrapping the code in a main function, adding some print statements, compiling it and running it.
+
+```cpp
+#include <iomanip>
+#include <iostream>
+
+#include "newtonraphson.hpp"
+
+int main() {
+  double initial_guess = -4;
+  double tolerance = 0.001;
+  NewtonRaphson finder(tolerance);
+  double x1 = finder.solve(initial_guess);
+
+  std::cout << std::fixed;
+  std::cout << std::setprecision(2);
+
+  for (Iteration iteration : finder.iterations) {
+    std::cout << "index = " << iteration.index;
+    std::cout << " x = " << iteration.x;
+    std::cout << " y = " << iteration.y;
+    std::cout << " slope = " << iteration.slope;
+    std::cout << " delta_x = " << iteration.delta_x << std::endl;
+  }
+  return 0;
+}
+```
+File: _cli.cpp_.
+
+Compile it with
+
+```shell
+c++ -o newtonraphson -I. newtonraphson.cpp cli.cpp
+```
+
+Run with
+
+```shell
+./newtonraphson
+i = 0 x = -4.00 y = -78.00 slope = 56.00 delta_x = -1.39
+i = 1 x = -2.61 y = -22.52 slope = 25.61 delta_x = -0.88
+i = 2 x = -1.73 y = -6.14 slope = 12.41 delta_x = -0.49
+i = 3 x = -1.23 y = -1.39 slope = 7.02 delta_x = -0.20
+i = 4 x = -1.03 y = -0.18 slope = 5.28 delta_x = -0.03
+i = 5 x = -1.00 y = -0.00 slope = 5.01 delta_x = -0.00
+```
+
+The last iteration has `-1.00` as x, which is what we expected.
+
+## Bindings
+
+Emscripten can handle simple types like double and int, but needs help exposing more complex types to JavaScript like the iterations property.
+We need to use [value_object](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#value-types) to expose the Iteration struct and [register_vector](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#built-in-type-conversions) as the iterations property type.
+
+So the bindings look like
+
+```cpp
+#include <emscripten/bind.h>
+#include "newtonraphson.hpp"
+
+using namespace emscripten;
+
+EMSCRIPTEN_BINDINGS(newtonraphson) {
+  class_<NewtonRaphson>("NewtonRaphson")
+    .constructor<double>()
+    .function("solve", &NewtonRaphson::solve)
+    .property("iterations", &NewtonRaphson::iterations)
+    ;
+
+  value_object<Iteration>("Iteration")
+    .field("index", &Iteration::index)
+    .field("x", &Iteration::x)
+    .field("y", &Iteration::y)
+    .field("slope", &Iteration::slope)
+    .field("delta_x", &Iteration::delta_x)
+    ;
+
+  register_vector<Iteration>("vector<Iteration>");
+}
+
+```
+File: _bindings.cpp_.
+
+Same as in previous blog we can compile to a WebAssembly module with Emscripten using `emcc` command
+
+```shell
```

Comment:
This was updated in the first blog post to have 2 lines.

Reply:
Done